src/algebra/homology/local_cohomology.lean

```diff
@@ -0,0 +1,181 @@
+variables {R : Type max u v} [comm_ring R] {D : Type v} [small_category D]
+
+/-- `local_cohomology.of_diagram I i` is the the functor sending an module `M` over a commutative
+ring `R` to the direct limit of `Ext^i(R/J, M)`, where `J` ranges over a collection of ideals
+of `R`, represented as a functor `I`. -/
+/-
+In this definition we do not assume any special property of the diagram `I`, but the relevant case
+will be where `I` is (cofinal with) the diagram of powers of a single given ideal.
+
+Below, we give two equivalent (to be shown) definitions of the usual local cohomology with support
+in an ideal `J`, `local_cohomology` and `local_cohomology.of_radical_eq_radical`.
+
+TODO: Show that any functor cofinal with `I` gives the same result.
+ -/
+def of_diagram (I : D ⥤ ideal R) (i : ℕ) :
+  Module.{max u v} R ⥤ Module.{max u v} R :=
+colimit (diagram.{(max u v) v} I i)
+
+end
+
+section diagrams
+
+variables {R : Type u} [comm_ring R]
+
+/-- The functor sending a natural number `i` to the `i`-th power of the ideal `J` -/
+def ideal_powers (J : ideal R) : ℕᵒᵖ ⥤ ideal R :=
+{ obj := λ t, J^(unop t),
+  map := λ s t w, ⟨⟨ideal.pow_le_pow w.unop.down.down⟩⟩, }
+
+/-- The full subcategory of all ideals with the same radical as a given ideal `J` -/
+@[reducible] def ideals_with_same_radical (J : ideal R) : Type u :=
+full_subcategory (λ J' : ideal R, J.radical ≤ J'.radical)
+
+/-- The diagram of all ideals with the same radical as `J`, represented as a functor.
+This is the "largest" diagram that computes local cohomology with support in `J`. -/
+def same_radical_diagram (J : ideal R) : (ideals_with_same_radical J) ⥤ ideal R :=
+full_subcategory_inclusion _
+
+end diagrams
+
+end local_cohomology
+
+/- We give two models for the local cohomology with support in an ideal `J`: first in terms of
+the powers of `J` (`local_cohomology`), then in terms of all ideals with the same radical
+as `J` (`local_cohomology.of_radical_eq_radical`). -/
+section models_for_local_cohomology
+
+open local_cohomology
+
+variables {R : Type u} [comm_ring R]
+
+/-- `local_cohomology J i` is `i`-th the local cohomology module of a module `M` over
+a commutative ring `R` with support in the ideal `J` of `R`, defined as the direct limit
+of Ext(R/J^t, M) over all powers `t`. -/
+def local_cohomology (J : ideal R) (i : ℕ) : Module.{u} R ⥤ Module.{u} R :=
+  of_diagram (ideal_powers J) i
+
+/-- Local cohomology as the direct limit of Ext(R/J, M) over *all* ideals with the same radical
+as `J`. -/
+def local_cohomology.of_radical_eq_radical (J : ideal R) (i : ℕ) :
+  Module.{u} R ⥤ Module.{u} R :=
+of_diagram.{u} (same_radical_diagram.{u} J) i
+-- TODO: Construct `local_cohomology J i ≅ local_cohomology.of_radical_eq_radical J i`
+
+end models_for_local_cohomology
+
+
+section local_cohomology_equiv
+
+open local_cohomology
+
+variables {R : Type u} [comm_ring R] (I J : ideal R)
+
+/-- Lifting the `ideal_powers J` diagram from a diagram valued in `ideals R` to a diagram
```

Comment:
OK I am almost completely happy with this PR. My only concern is that these last three declarations are in the root namespace. Maybe put them in `ring_theory` or something? Maybe take a look at which namespace the various theorems and definitions about radical ideals are and put them there? Especially that last one.

Reply:
Okay, I put the two into `local_cohomology` and the algebra one into the `ideal` namespace.